Core/src/Testing/DatastoreOperationRefreshTrait.php

```diff
@@ -57,13 +60,87 @@ public function refreshOperation($stub, ConnectionInterface $connection, array $
             $options['returnInt64AsObject']
         );
 
+        $serializer = $this->getSerializer();
+
         $stub->___setProperty('operation', new Operation(
-            $connection,
+            $requestHandler,
+            $serializer,
             $options['projectId'],
             $options['returnInt64AsObject'],
             $mapper
         ));
 
         return $stub;
     }
+
+    /**
+     * Helper method for Unit and Snippet test classes. This mocks the
+     * $requestHandler class property present in the Test Class with
+     * given arguments.
+     *
+     * @param string $methodName The method name to mock in RequestHandler::sendRequest
+     * @param array<string, mixed> $params The parameters to look for in the
+     * array equivalent of rpc request.
+     * @param mixed $returnValue The value to be returned by sendRequest mock.
+     * @param null|int $shouldBeCalledTimes Adds a shouldBeCalled prophecy. Defaults to `null`, implying nothing is added.
+     * [
+     *      `0` => `shouldBeCalled`,
+     *      Non zero positive integer $x => `shouldBeCalledTimes($x)`
+     * ]
+     */
+    private function mockSendRequest($methodName, $params, $returnValue, $shouldBeCalledTimes = null)
+    {
+        $serializer = $this->getSerializer();
+
+        $prophecy = $this->requestHandler->sendRequest(
+            DatastoreClient::class,
+            $methodName,
+            Argument::that(function ($arg) use ($methodName, $params, $serializer) {
+                $requestName = ucfirst($methodName . 'Request');
+                $x = explode('\\', get_class($arg));
+                $argName = end($x);
+
+                if ($requestName != $argName) {
+                    return false;
+                }
+                $data = $serializer->encodeMessage($arg);
+                return array_replace_recursive($data, $params) == $data;
+            }),
+            Argument::cetera()
+        );
+
+        if (!is_null($shouldBeCalledTimes)) {
+            if ($shouldBeCalledTimes == 0) {
+                $prophecy->shouldBeCalled();
+            } else {
+                $prophecy->shouldBeCalledTimes($shouldBeCalledTimes);
+            }
+        }
+
+        $prophecy->willReturn($returnValue);
+    }
+
+    private function getSerializer()
+    {
+        if (isset($this->serializer)) {
```

Comment:
I'm confused because the property `$serializer` is not defined in the trait, and `$this->serializer` is never set as a result of the function. Is this supposed to be something like this?
```php
        if (!isset($this->serializer)) {
            $this->serializer = new Serializer(...);
        }

        return $this->serializer;
```

Reply:
Now I see we have a static `self::$_serializer` _and_ an instance `$this->serializer`. This is even more confusing, IMHO. Why would we have both? Where does `$this->serializer` come from?